src/middlewares/restrict-user-details.js

```diff
@@ -0,0 +1,114 @@
+/***
+ * Main algorithm here: take the response and introspect all attributes, if any attribute is a relation, we take that object and introspect its attributes. 
+ * When relation points to a user or profile, we replace that object with shrunk version, including only allowed fields.
+ * Using private fields in Strapi Schema is not an option, as they are also not available for the user calling `/api/users/me` or `/api/users/:selfId`.
+ * The goal of this middleware is to restrict those fields for non-owner users.
+ * 
+ * Metadata is not properly documented in Strapi. 
+ * It is a Mapa Map-alike object where key is a modelName and value is MetaModel, which contains almost all the same info you can find under `content-types/schema.json`. 
+ * See https://github.com/strapi/strapi/blob/develop/packages/core/database/src/types/index.ts#L216 and https://github.com/strapi/strapi/blob/develop/packages/core/database/src/metadata/metadata.ts
+ */
+
+module.exports = (config, { strapi }) => {
+  return async (ctx, next) => {
+    await next();
+    try {
+      // Only need to filter if there is a response body
+      if (ctx.request.method === "GET" && ctx.request.url.includes("/api/") && ctx.response.body) {
```

Comment:
@becevka I added a check for non empty `ctx.response.body`, otherwise https://github.com/dev-launchers/strapiv4/pull/196/files#diff-020f47be16701da9ac1d16017015d4d6d586e314e61850aa215c4c801450563eR27 will set it to an empty body and change the status code from 404 to 204.